File: src/execgo_playground/benchmarks/runner.py
from __future__ import annotations

from pathlib import Path

from ..models import BenchmarkResult, BenchmarkRunRequest
from ..runner import ExperimentRunner
from ..utils import write_json, write_text
from .report import render_benchmark_summary


class BenchmarkRunner:
    def __init__(self, runner: ExperimentRunner) -> None:
        self.runner = runner

    def run(self, request: BenchmarkRunRequest) -> list[BenchmarkResult]:
        results: list[BenchmarkResult] = []
        for framework in request.frameworks:
            for scenario in request.scenarios:
                for chaos_profile in request.chaos_profiles:
                    for repetition in range(request.repetitions):
                        result = self.runner.run(
                            framework=framework,
                            scenario_id=scenario,
                            mode=request.mode,
                            chaos_profile_id=chaos_profile,
                            model_profile=request.model_profile,
                            repetition=repetition,
                        )
                        results.append(result)
        summary_path = Path(self.runner.harness.run_root) / "benchmark-summary.json"
        write_json(summary_path, [result.model_dump(mode="json") for result in results])
        write_text(summary_path.with_suffix(".md"), render_benchmark_summary(results))
        return results

File: src/execgo_playground/benchmarks/runner.ts
import path from "node:path";
import type { BenchmarkResult, BenchmarkRunRequest } from "../models.js";
import { ExperimentRunner } from "../runner.js";
import { writeJson, writeText } from "../utils.js";
import { renderBenchmarkSummary } from "./report.js";

export class BenchmarkRunner {
  constructor(private readonly runner: ExperimentRunner) {}

  async run(request: BenchmarkRunRequest): Promise<BenchmarkResult[]> {
    const results: BenchmarkResult[] = [];
    for (const framework of request.frameworks) {
      for (const scenario of request.scenarios) {
        for (const chaosProfile of request.chaos_profiles) {
          for (let repetition = 0; repetition < request.repetitions; repetition += 1) {
            const result = await this.runner.run({
              framework,
              scenarioId: scenario,
              mode: request.mode,
              chaosProfileId: chaosProfile,
              modelProfile: request.model_profile,
              repetition,
            });
            results.push(result);
          }
        }
      }
    }
    const summaryPath = path.join(this.runner.harness.runRoot, "benchmark-summary.json");
    await writeJson(summaryPath, results);
    await writeText(summaryPath.replace(/\.json$/, ".md"), renderBenchmarkSummary(results));
    return results;
  }
}
